### PR TITLE
Add optional publishing rate limit

### PR DESCRIPTION
**Pull request by Friede80**
This adds an option to limit the rate of publishing. When a rate limit is provided, publishing is done serially instead of concurrently. The period between sending packets is calculated as the `maxPacketSize / rateLimit` since any message smaller than `maxPacketSize` can be sent in one packet and thus can't be rate limited.

### CODE REVIEW

File: src/Network/GRPC/MQTT/Message/Packet.hs [hunk 1]
```diff
@@ -208,37 +211,58 @@ maxPacketSize = 256 * 2 ^ (20 :: Int) - 128 * 2 ^ (10 :: Int)
...
+            Nothing -> pure ()
+            Just jobid -> do
+              let packet = makePacket jobid
+              publish (wireWrapPacket' packet)
+              next
+      Just rateLimit -> do
+        -- Throughput is rate limited on a 1 second interval and we don't subdivide packets to comply
+        -- with the rate limit, so the payload size must be less than or equal to the rate limit.
+        -- NB: Since `maxPayloadSize` is already at least 1, this ensures that `rateLimit` is not 0.
+        unless (maxPayloadSize <= rateLimit) $
```
Comment (ixmatus) [inline src/Network/GRPC/MQTT/Message/Packet.hs:249]: Can you push this further upstream? e.g if `rateLimit` were a `Natural` then we wouldn't need this check.

Comment (ixmatus) [reply src/Network/GRPC/MQTT/Message/Packet.hs:249]: Ah, I thought that might be the case but like you said I also think it's still a better fit if we never need the value to be negative or zero.

File: src/Network/GRPC/MQTT/Message/Packet.hs [hunk 2]
```diff
@@ -208,37 +211,58 @@ maxPacketSize = 256 * 2 ^ (20 :: Int) - 128 * 2 ^ (10 :: Int)
...
+          throwIO (PacketSizeError $ toInteger packetSizeLimit)
+
+        -- Convert `maxPayloadSize` and `rateLimit` to Integer first to avoid overflow. Also,
+        -- the above condition that `maxPayloadSize` <= `rateLimit` ensures that the period is
+        -- between 0 and 1,000,000 microseconds, so the final conversion Integer -> Int is always safe.
+        let packetSendPeriod :: Int
+            packetSendPeriod = fromIntegral @Integer @Int $ fromIntegral maxPayloadSize * 1_000_000 `div` fromIntegral rateLimit
+
+        forM_ [0 .. njobs] $ \jobid -> do
+          timer <- async $ threadDelay packetSendPeriod
```
Comment (riz0id) [inline src/Network/GRPC/MQTT/Message/Packet.hs:259]: Could this be accomplished with [`TSem`](https://hackage.haskell.org/package/stm-2.5.1.0/docs/Control-Concurrent-STM-TSem.html#t:TSem)? You could have a separate thread running thats loop `packetSendPeriod` that is bound by the total number of bytes to be sent. This loop would just accumulate the number of bytes that can be published into the [`TSem`](https://hackage.haskell.org/package/stm-2.5.1.0/docs/Control-Concurrent-STM-TSem.html#t:TSem). Then, in the loop running on a separate thread handling the packet publishes here

```haskell
        replicateConcurrently_ caps $ fix \next ->
          atomically (takeJobId jobs) >>= \case
            Nothing -> pure ()
            Just jobid -> do
              let packet = makePacket jobid
              publish (wireWrapPacket' packet)
              next
```

and a function which can take `n` bytes bounded by `maxPayloadSize`, roughly something like:

```haskell
waitTSemN :: TSem () -> Integer -> STM ()
waitTSemN (TSem sem) limit = do 
  i <- readTVar sem 
  when (i <= 0) retry
  let ret = min limit i
  writeTVar sem (i - ret)
  pure ret
```

and incorporate into the packet publish with something like `makePacket jobid =<< atomically takeTSemN`. The loop accumulating into the rate-limiting `TSem` and the packet publishing loop could be run in parallel with `concurrently_`.

I think the big advantage to handling throttling this way is that the logic for the throttling and the logic for sending packets is kept disjoint, removing the additional case we're handling here with `Maybe`.

Comment (Friede80) [reply src/Network/GRPC/MQTT/Message/Packet.hs:257]: I like the idea of separating the throttling and sending logic, but I'm not quite following your suggestion here using `TSem`. Semaphores are really just counters. You can't read/write data to them.

**Review by ixmatus — APPROVED** [review]
I also like the suggestion to decouple the rate limiting logic from the sending logic but don't have any specific suggestions.

File: src/Network/GRPC/MQTT/Message/Packet.hs (hunk 2 above)
Comment (riz0id) [reply src/Network/GRPC/MQTT/Message/Packet.hs:257]: Yeah I was just providing a rough idea, maybe `TSem` specifically doesn't make sense in this case. The core idea I was trying to convey was just to have a `TVar Int` that carries the number of bytes that can be sent before the packet sender gets throttled, i.e. if `0 <- readTVar (var :: TVar Int)` then the packet sender blocks until `readTVar var > 0`. A separate thread could add some number of bytes to `var :: TVar Int` every `packetSendPeriod` to unblock the publishing thread until all bytes are sent.

**Review by riz0id — APPROVED** [review]

**Pull request merged**